File: app/src/main/java/com/acceleratorer/wuwavn/Sha256Verifier.java
package com.acceleratorer.wuwavn;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.security.MessageDigest;

public final class Sha256Verifier {
    private Sha256Verifier() {
    }

    public static String sha256(File file) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        try (InputStream input = new FileInputStream(file)) {
            byte[] buffer = new byte[8192];
            int read;
            while ((read = input.read(buffer)) != -1) {
                digest.update(buffer, 0, read);
            }
        }

        byte[] hash = digest.digest();
        StringBuilder hex = new StringBuilder(hash.length * 2);
        for (byte b : hash) {
            hex.append(String.format("%02x", b));
        }
        return hex.toString();
    }

    public static boolean verify(File file, String expectedSha256) throws Exception {
        return sha256(file).equalsIgnoreCase(expectedSha256);
    }
}

File: app/src/main/kotlin/com/acceleratorer/wuwavn/Sha256Verifier.kt
package com.acceleratorer.wuwavn

import java.io.File
import java.io.FileInputStream
import java.security.MessageDigest

object Sha256Verifier {
    fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        FileInputStream(file).use { input ->
            val buffer = ByteArray(8192)
            while (true) {
                val read = input.read(buffer)
                if (read == -1) break
                digest.update(buffer, 0, read)
            }
        }

        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    fun verify(file: File, expectedSha256: String): Boolean =
        sha256(file).equals(expectedSha256, ignoreCase = true)
}
